Migrate nonzero from the TH to Aten(CPU)

ifedan opened the pull request:
#24745

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Merge branch 'master' into nonzero

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Merge branch 'master' into nonzero

ifedan committed:
Merge branch 'nonzero' of https://github.com/ifedan/pytorch into nonzero

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Merge branch 'master' into nonzero

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Revert "Migrate nonzero from the TH to Aten(CPU)"

This reverts commit 8f6a9dd69fdc45e4f5c8ecb851aee5c65bfa3580.

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan commented:
![image](https://user-images.githubusercontent.com/12058312/66356832-86c5dd80-e93a-11e9-9b88-2a9b1dfb46f9.png)

ifedan commented:
```python
import torch
import timeit
import itertools
import statistics

for i in [10, 100, 10000, 100000, 1000000]:
    setup_code = '''import torch; w = (torch.rand({0}) < 0.5).double()'''.format(i)
    test_code = '''torch.nonzero(w);'''
    result = timeit.repeat(setup = setup_code,stmt = test_code,repeat = 3,number = 100)
    mean = statistics.mean(result)
    std = statistics.stdev(result)
    print('''size = {0} time = {1}'''.format(i, mean))
```

ifedan commented:
https://github.com/pytorch/pytorch/pull/27513

VitalyFedyunin reviewed (changes requested):
Not working for non contiguous tensors.

```python 

In [8]: x = (torch.randn(2,5)<0.5).double()                                                                                                                                                                                                                                                

In [9]: x.t()                                                                                                                                                                                                                                                                              
Out[9]: 
tensor([[1., 1.],
        [1., 1.],
        [1., 0.],
        [0., 0.],
        [1., 1.]], dtype=torch.float64)

In [10]: x                                                                                                                                                                                                                                                                                 
Out[10]: 
tensor([[1., 1., 1., 0., 1.],
        [1., 1., 0., 0., 1.]], dtype=torch.float64)

In [11]: x.t().is_contiguous()                                                                                                                                                                                                                                                             
Out[11]: False

In [12]: x.t().nonzero()                                                                                                                                                                                                                                                                   
Out[12]: 
tensor([[0, 0],
        [0, 1],
        [1, 0],
        [2, 0],
        [2, 1],
        [3, 0],
        [4, 1]])

In [13]: x.t().nonzero()                                                                                                                                                                                                                                                                   
Out[13]: 
tensor([[0, 0],
        [0, 1],
        [1, 0],
        [2, 0],
        [2, 1],
        [3, 0],
        [4, 1]])

In [14]: x.t().contiguous().nonzero()                                                                                                                                                                                                                                                      
Out[14]: 
tensor([[0, 0],
        [0, 1],
        [1, 0],
        [1, 1],
        [2, 0],
        [4, 0],
        [4, 1]])
                                                                                                                                                                                                                                                                         
```

compare last 2 statements.

As `cpu_serial_kernel` doesn't guarantee traverse order.

albanD commented:
Can the TH implementation be deleted? Or is it used in other places?

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Removed TH implementation

ifedan committed:
Fix tests

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan commented:
> Can the TH implementation be deleted? Or is it used in other places?

deleted

ifedan commented:
> Not working for non contiguous tensors.
> 
> ```python
> In [8]: x = (torch.randn(2,5)<0.5).double()                                                                                                                                                                                                                                                
> 
> In [9]: x.t()                                                                                                                                                                                                                                                                              
> Out[9]: 
> tensor([[1., 1.],
>         [1., 1.],
>         [1., 0.],
>         [0., 0.],
>         [1., 1.]], dtype=torch.float64)
> 
> In [10]: x                                                                                                                                                                                                                                                                                 
> Out[10]: 
> tensor([[1., 1., 1., 0., 1.],
>         [1., 1., 0., 0., 1.]], dtype=torch.float64)
> 
> In [11]: x.t().is_contiguous()                                                                                                                                                                                                                                                             
> Out[11]: False
> 
> In [12]: x.t().nonzero()                                                                                                                                                                                                                                                                   
> Out[12]: 
> tensor([[0, 0],
>         [0, 1],
>         [1, 0],
>         [2, 0],
>         [2, 1],
>         [3, 0],
>         [4, 1]])
> 
> In [13]: x.t().nonzero()                                                                                                                                                                                                                                                                   
> Out[13]: 
> tensor([[0, 0],
>         [0, 1],
>         [1, 0],
>         [2, 0],
>         [2, 1],
>         [3, 0],
>         [4, 1]])
> 
> In [14]: x.t().contiguous().nonzero()                                                                                                                                                                                                                                                      
> Out[14]: 
> tensor([[0, 0],
>         [0, 1],
>         [1, 0],
>         [1, 1],
>         [2, 0],
>         [4, 0],
>         [4, 1]])
>                                                                                                                                                                                                                                                                          
> ```
> 
> compare last 2 statements.
> 
> As `cpu_serial_kernel` doesn't guarantee traverse order.

fixed and added tests

ifedan commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -353,6 +357,7 @@ struct CAFFE2_API TensorIterator {
   bool promote_gpu_output_dtypes_ = false;
   bool final_output_ = true;
   bool check_mem_overlap_ = false;
+  bool reverse_order_dims_ = false;
```
maybe we need to use a better name

ifedan committed:
Merge branch 'master' into nonzero

ifedan committed:
Migrate nonzero from the TH to Aten(CPU)

ifedan committed:
Merge branch 'master' into nonzero

albanD commented:
TH is no more so this already happened somewhere else I'm sure.
